control/lti.py

```diff
@@ -47,10 +47,47 @@ def __init__(self, inputs=1, outputs=1, dt=None):
         """Assign the LTI object's numbers of inputs and ouputs."""
 
         # Data members common to StateSpace and TransferFunction.
-        self.inputs = inputs
-        self.outputs = outputs
+        self.ninputs = inputs
+        self.noutputs = outputs
         self.dt = dt
 
+    #
+    # Getter and setter functions for legacy input/output attributes
+    #
+    # For this iteration, generate a pending deprecation warning whenever
+    # the getter/setter is called.  For a future iteration, turn it into a
+    # deprecation warning.
+    #
+
+    @property
+    def inputs(self):
+        raise PendingDeprecationWarning(
+            "The LTI `inputs` attribute will be deprecated in a future "
+            "release.  Use `ninputs` instead.")
+        return self.ninputs
```

Comment:
This raises an Exception of the class `PendingDeprecationWarning` and never returns self.inputs.

```
Python 3.8.6 (default, Nov 09 2020, 12:09:06) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import control
>>> rsys = control.rss(3,3,3)
>>> ninputs = rsys.inputs
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ben/src/python-control/control/lti.py", line 64, in inputs
    raise PendingDeprecationWarning(
PendingDeprecationWarning: The LTI `inputs` attribute will be deprecated in a future release.  Use `ninputs` instead.
>>> ninputs
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'ninputs' is not defined
```

```suggestion
        warn("The LTI `inputs` attribute will be deprecated in a future "
             "release.  Use `ninputs` instead.",
             PendingDeprecationWarning, stacklevel=2)
        return self.ninputs
```

Additionally, either a warning filter needs to be set in order to make the warning visible to end users, or a different category than `*DeprecationWarning` needs to be used.

```python
filterwarnings('module',
               message=".*LTI.*attribute.*deprecated",
               category=PendingDeprecationWarning)
```

```
Python 3.8.6 (default, Nov 09 2020, 12:09:06) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import control; rsys = control.rss(3,3,3); ninputs = rsys.inputs
<stdin>:1: PendingDeprecationWarning: The LTI `inputs` attribute will be deprecated in a future release.  Use `ninputs` instead.
>>> ninputs
3
```

- Please add a check using pytest.warns().

Reply:
I've fixed these up and decide to use DeprecationWarning for now (keep things internal to developers) but in a future release (0.9.x) we should change to FutureWarning, which will make it show up for users as well.

Reply:
Still using` raise`  instead of `warning`. Now we are at:

> This raises an Exception of the class ` DeprecationWarning` and never returns self.inputs.

```
Python 3.8.6 (default, Nov 09 2020, 12:09:06) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import control; rsys = control.rss(3,3,3); ninputs = rsys.inputs
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ben/src/python-control/control/lti.py", line 64, in inputs
    raise DeprecationWarning(
DeprecationWarning: The LTI `inputs` attribute will be deprecated in a future release.  Use `ninputs` instead.
>>> ninputs
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'ninputs' is not defined
>>> 
```